SEMANTICS.md

```diff
@@ -0,0 +1,297 @@
+pub enum ReplyOn {
+    /// Always perform a callback after SubMsg is processed
+    Always,
+    /// Only callback if SubMsg returned an error, no callback on success case
+    Error,
+    /// Only callback if SubMsg was successful, no callback on error case
+    Success,
+}
+```
+
+What are the semantics of a submessage execution. First, we create a
+sub-transaction context around the state, allowing it to read the latest state
+written by the caller, but write to yet-another cache. If `gas_limit` is set, it
+is sandboxed to how much gas it can use until it aborts with `OutOfGasError`.
+This error is caught and returned to the caller like any other error returned
+from contract execution (unless it burned the entire gas limit of the
+transaction). What is more interesting is what happens on completion.
+
+If it return success, the temporary state is committed (into the caller's
+cache), and the `Response` is processed as normal (an event is added to the
+current EventManager, messages and submessages are executed). Once the
+`Response` is fully processed, this may then be intercepted by the calling
+contract (for `ReplyOn::Always` and `ReplyOn::Success`). On an error, the the
+subcall will revert any partial state changes due to this message, but not
+revert any state changes in the calling contract. The error may then be
+intercepted by the calling contract (for `ReplyOn::Always` and
+`ReplyOn::Error`). _In this case, the messages error doesn't abort the whole
+transaction_
+
+#### Handling the Reply
+
+In order to make use of `submessages`, the calling contract must have an extra
+entry point:
+
+```rust
+#[entry_point]
+pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractError> { }
+
+pub struct Reply {
+    pub id: u64,
+    /// ContractResult is just a nicely serializable version of `Result<T, String>`
+    pub result: ContractResult<SubcallResponse>,
+}
+
+pub struct SubcallResponse {
+    pub events: Vec<Event>,
+    pub data: Option<Binary>,
+}
+```
+
+After the `submessage` is finished, the caller will get a chance to handle the
+result. It will get the original `id` of the subcall so it can switch on how to
+process this, and the `Result` of the execution, both success and error. Note
+that it includes all events returned by the submessage, which applies to native
+sdk modules as well (like Bank) as well as the data returned from below. This
+and the original call id provide all context to continue processing it. If you
+need more state, you must save some local context to the store (under the `id`)
+before returning the `submessage` in the original `execute`, and load it in
+`reply`. We explicitly prohibit passing information in contract memory, as that
+is the key vector for reentrancy attacks, which are a large security surface
+area in Ethereum.
+
+The `reply` call may return `Err` itself, in which case it is treated like the
+caller errored, and aborting the transaction. (Unless, of course, the original
```

Comment:
> (Unless, of course, the original...
The sentence in brackets make it hard for me to follow the flow. Let's get the reply flow described first and explain this at the end of the paragraph? 
Maybe with an example:

Please keep in mind that `submessage` `execution` and `reply` can happen within the context of another `submessage`. For example `contract-A --submessage--> contract-B --submessage->contract-C`. 
Then `contract-B` can revert the state for `contract-C` and itself by returning an error in the submessage `reply` handling. `contract-A` can revert state for itself, `contract-B` and `contract-C` with it's submessage `reply` error result..